### PR TITLE
EDSC-4151: Convert /path/search/granules.cy.js to Playwright

### PR DESCRIPTION
**Pull request by dpesall**
# Overview

### What is the feature?

Converting Granules cypress tests to playwright

### What is the Solution?

Rewrite relevant cypress tests in playwright.

### What areas of the application does this impact?

e2e testing

# Testing

### Reproduction steps

1. Run playwright tests in the following path: `paths/search/granules/granules.spec.js`
2. Verify that all tests pass

# Checklist

- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 1]
```diff
@@ -0,0 +1,1263 @@
...
+        })
+      })
+
+      await page.goto('/search/granules?p=C1214470488-ASF&pg[0][bo]=true')
+
+      // Ensure the correct number of results were loaded
+      await testResultsSize(page, cmrHits)
+
+      // Checkboxes are checked correctly
+      await expect(page.locator('[data-testid="granule-filters__browse-only"]')).toBeChecked()
```
Comment (eudoroolivares2016) [inline tests/e2e/paths/search/granules/granules.spec.js:441]: use getBytestId for these two asserts and the one below

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 2]
```diff
@@ -0,0 +1,1263 @@
...
+
+      await page.route('**/api', (route) => {
+        const body = JSON.stringify(route.request().postData())
+
+        if (body === JSON.stringify(graphQlGetCollection(conceptId))) {
+          route.fulfill({
+            body: JSON.stringify(focusedGranuleCollectionGraphQlBody),
+            headers: focusedGranuleGraphQlHeaders
+          })
+        } else if (body === '{"query":"\\n    query GetGranule(\\n      $params: GranuleInput\\n    ) {\\n      granule(\\n        params: $params\\n      ) {\\n        granuleUr\\n        granuleSize\\n        title\\n        onlineAccessFlag\\n        dayNightFlag\\n        timeStart\\n        timeEnd\\n        dataCenter\\n        originalFormat\\n        conceptId\\n        collectionConceptId\\n        spatialExtent\\n        temporalExtent\\n        relatedUrls\\n        dataGranule\\n        measuredParameters\\n        providerDates\\n      }\\n    }","variables":{"params":{"conceptId":"G2058417402-LPDAAC_ECS"}}}') {
```
Comment (eudoroolivares2016) [inline tests/e2e/paths/search/granules/granules.spec.js:1021]: This string feels like it should be a separate file but, I know it isn't on the existing cypress test and I we already have alot of mocks so I'm fine either way

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 3]
```diff
@@ -0,0 +1,1263 @@
...
+      const cmrHits = 1
+
+      await page.route('**/search/granules.json', (route) => {
+        const request = route.request()
+        const body = request.postData()
+        expect(body).toBe('echo_collection_id=C1214470488-ASF&options[readable_granule_name][pattern]=true&page_num=1&page_size=20&readable_granule_name[]=S1A_S3_SLC__1SDH_20140615T034444_20140615T034512_001055_00107C_16F1')
+        route.fulfill({
+          body: JSON.stringify(readableGranuleNameGranulesBody),
+          headers: {
+            ...commonGranulesHeaders,
```
[the remark this answers is not in the document]
Comment (dpesall) [reply tests/e2e/paths/search/granules/granules.spec.js:217]: I'm not sure either, I just following the methodology from the Cypress tests

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 4]
```diff
@@ -0,0 +1,1263 @@
...
+        const cmrHits = 17231
+
+        await page.route('**/search/granules.json', (route) => {
+          const request = route.request()
+          const body = request.postData()
+          expect(body).toBe('echo_collection_id=C1214470488-ASF&page_num=1&page_size=20&temporal=2020-01-01T00:00:00.000Z,2020-01-31T23:59:59.999Z')
+          route.fulfill({
+            body: JSON.stringify(temporalGranulesBody),
+            headers: {
+              ...commonGranulesHeaders,
```
[the remark this answers is not in the document]
Comment (dpesall) [reply tests/e2e/paths/search/granules/granules.spec.js:275]: See above comment

File: tests/support/graphQlGetCollections.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+/**
+ * Return the GraphQl query for verifying the request is what we expect inside of cy.intercept
```
Comment (eudoroolivares2016) [inline tests/support/graphQlGetCollections.js:2]: Gotcha just fix the doc-string

```suggestion
 * Return the GraphQl query for verifying the request is what we expect inside of page.route
```

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 6]
```diff
@@ -0,0 +1,1244 @@
...
+    .getByTestId('panel-group-header__heading-meta-text')
+    .textContent()
+  expect(metaText).toBe(`Showing ${expectedSize} of ${commafy(cmrHits)} matching ${pluralize('granule', cmrHits)}`)
+}
+
+test.describe('Path /search/granules', () => {
+  test.beforeEach(async ({ page }) => {
+    const fakeNow = new Date('2021-06-01').valueOf()
+
+    await page.addInitScript(`{
```
Comment (eudoroolivares2016) [inline tests/e2e/paths/search/granules/granules.spec.js:124]: This feels like it does not belong here... why is this needed now it wasn't prior to the last commits

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 7]
```diff
@@ -0,0 +1,1244 @@
...
+        }
+
+        route.fulfill({
+          body: JSON.stringify(noParamsTimelineBody),
+          headers: noParamsTimelineHeaders
+        })
+      })
+
+      await page.route('**/api', (route) => {
+        const request = route.request()
```
Comment (eudoroolivares2016) [inline tests/e2e/paths/search/granules/granules.spec.js:174]: I don't think we need to do the stringify and unstringify stuff here I think a few other spots correct me if Im wrong

File: tests/e2e/paths/search/granules/granules.spec.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (eudoroolivares2016) [reply tests/e2e/paths/search/granules/granules.spec.js:210]: Okay since this is quite large we can just fix that once we're all merged in

File: tests/e2e/paths/search/granules/granules.spec.js (hunk 6 above)
Comment (dpesall) [reply tests/e2e/paths/search/granules/granules.spec.js:124]: I realized there was a bug with the previous date code setting code, and it didn't actually work for setting the month properly. It was fortunate that today turned to August and caused the bug to show up. This was the best workaround I could find: https://betterstack.com/community/questions/playwright-mock-date/

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 8]
```diff
@@ -0,0 +1,1180 @@
...
+import timelineGraphQlBody from './__mocks__/timeline/graphql.body.json'
+import timelineGraphQlHeaders from './__mocks__/timeline/graphql.headers.json'
+import timelineTimelineBody from './__mocks__/timeline/timeline.body.json'
+import timelineTimelineHeaders from './__mocks__/timeline/timeline.headers.json'
+import graphQlHeaders from './__mocks__/common/graphql.headers.json'
+
+import { login } from '../../../../support/login'
+
+const defaultCmrPageSize = 20
+
```
Comment (eudoroolivares2016) [inline tests/e2e/paths/search/granules/granules.spec.js:110]:
```suggestion
/**
 * Tests the search panel header and meta text for results size
 * @param {Page} page Playwright page object
 * @param {Integer} cmrHits Total number of collections that match the query
 */
```

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 9]
```diff
@@ -0,0 +1,1180 @@
...
+          }
+
+          route.fulfill({
+            body: JSON.stringify(temporalTimelineBody),
+            headers: temporalTimelineHeaders
+          })
+        })
+
+        await page.route('**/api', (route) => {
+          const request = route.request()
```
Comment (eudoroolivares2016) [inline tests/e2e/paths/search/granules/granules.spec.js:265]:
```suggestion
 const requestBody = JSON.parse(route.request().postData())
          const expectedBody = JSON.parse(graphQlGetCollection(conceptId))
          expect(requestBody).toEqual(expectedBody)```
          
       Can this one be like the others?

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 10]
```diff
@@ -0,0 +1,1180 @@
...
+        } else if (body.data.query === JSON.parse(graphQlGetCollection(conceptId)).query) {
+          await route.fulfill({
+            body: JSON.stringify(subscriptionGraphQlBody),
+            headers: graphQlHeaders
+          })
+        }
+      })
+
+      await page.goto('/search/granules?p=C1214470488-ASF&pg[0][gsk]=-start_date&sp[0]=-77.04119,38.80585')
+
```
Comment (eudoroolivares2016) [inline tests/e2e/paths/search/granules/granules.spec.js:1175]:
```suggestion
 // Ensure the correct number of results were loaded
      testResultsSize(page, cmrHits)
```
the cypress test had this assertion for this test

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 11]
```diff
@@ -0,0 +1,1180 @@
...
+          await route.fulfill({
+            body: JSON.stringify(projectGranuleProjectGranuleBody),
+            headers: {
+              ...commonGranulesHeaders,
+              'cmr-hits': '1'
+            }
+          })
+        }
+      })
+
```
Comment (eudoroolivares2016) [inline tests/e2e/paths/search/granules/granules.spec.js:1004]:
```suggestion
   await page.route(/saved_access_configs/, async (route) => {
        await route.fulfill({
          json: {}
        })
      })
```
In the cypress test we are mocking this request I'm not sure why we aren't here. I did pull it down and I was seeing a red banner warning about it

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 12]
```diff
@@ -0,0 +1,1180 @@
...
+        route.fulfill({
+          body: JSON.stringify(noParamsGranulesBody),
+          headers: {
+            ...commonGranulesHeaders,
+            'access-control-expose-headers': 'cmr-hits',
+            'cmr-hits': cmrHits
+          }
+        })
+      })
+
```
Comment (eudoroolivares2016) [inline tests/e2e/paths/search/granules/granules.spec.js:1075]:
```suggestion
 await page.route(/saved_access_configs/, async (route) => {
        await route.fulfill({
          json: {}
        })
      })
```
This is the other one where we were mocking this request on cypress same I saw the banner and then added this locally to make it go away

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 13]
```diff
@@ -0,0 +1,1180 @@
...
+              ...commonGranulesHeaders,
+              'access-control-expose-headers': 'cmr-hits',
+              'cmr-hits': cmrHits.toString()
+            }
+          })
+        } else if (body === 'echo_collection_id=C194001210-LPDAAC_ECS&page_num=1&page_size=1&concept_id[]=G2058417402-LPDAAC_ECS') {
+          await route.fulfill({
+            body: JSON.stringify(projectGranuleProjectGranuleBody),
+            headers: {
+              ...commonGranulesHeaders,
```
Comment (eudoroolivares2016) [inline tests/e2e/paths/search/granules/granules.spec.js:998]:
```suggestion
              ...commonGranulesHeaders,
              'access-control-expose-headers': 'cmr-hits',
```
I'm surprised we didn't need this here I think we should just add `'access-control-expose-headers': 'cmr-hits',`
to the `commonGranulesHeaders` file then you don't have ot append it each call

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 14]
```diff
@@ -0,0 +1,1180 @@
...
+
+        await page.route('**/search/granules/timeline', (route) => {
+          const request = route.request()
+          const body = request.postData()
+          if (body) {
+            expect(body).toBe('end_date=2023-12-01T00:00:00.000Z&interval=day&start_date=2018-12-01T00:00:00.000Z&concept_id[]=C1214470488-ASF')
+          }
+
+          route.fulfill({
+            body: JSON.stringify(temporalTimelineBody),
```
Comment (eudoroolivares2016) [inline tests/e2e/paths/search/granules/granules.spec.js:259]: Just in general the reason you have to use `Stringify` alot here can be worked around by doing a 
```suggestion
            json: temporalTimelineBody
```

File: tests/e2e/paths/search/granules/granules.spec.js (hunk 9 above)
Comment (eudoroolivares2016) [reply tests/e2e/paths/search/granules/granules.spec.js:265]: Can this one be like the others?

**Review by eudoroolivares2016 — APPROVED** [review]
Went through all the tests looks like it was covering the same stuff minor fixes I think will reduce some unnecessary calls

File: tests/e2e/paths/search/granules/granules.spec.js (hunk 7 above)
Comment (dpesall) [reply tests/e2e/paths/search/granules/granules.spec.js]: Simplified

File: tests/e2e/paths/search/granules/granules.spec.js (hunk 11 above)
Comment (dpesall) [reply tests/e2e/paths/search/granules/granules.spec.js:1004]: Added

File: tests/e2e/paths/search/granules/granules.spec.js (hunk 13 above)
Comment (dpesall) [reply tests/e2e/paths/search/granules/granules.spec.js:998]: I'm not sure either come to think of it, because this part of the conditional is definitely getting called. It may have been unnecessary in some of the tests but I just went through and added it to all of them (though I seemingly missed this one)

File: tests/e2e/paths/search/granules/__mocks__/temporal_recurring.body.json
Comment (trevorlang) [inline tests/e2e/paths/search/granules/__mocks__/temporal_recurring.body.json:1]: Using tabs instead of spaces

File: tests/e2e/paths/search/granules/__mocks__/tiling_system.body.json
Comment (trevorlang) [inline tests/e2e/paths/search/granules/__mocks__/tiling_system.body.json:1]: New line at end

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 15]
```diff
@@ -0,0 +1,1197 @@
...
+            'access-control-expose-headers': 'cmr-hits',
+            'cmr-hits': cmrHits
+          }
+        })
+      })
+
+      await page.route('**/search/granules/timeline', (route) => {
+        const request = route.request()
+        const body = request.postData()
+        if (body) {
```
Comment (trevorlang) [inline tests/e2e/paths/search/granules/granules.spec.js:492]: It seems kind of odd to have an if block in a test. I see this in a few spots. Why would body not exist when running this test?

File: tests/e2e/paths/search/granules/granules.spec.js [hunk 16]
```diff
@@ -0,0 +1,1197 @@
...
+
+  test.describe('When the path is loaded with browse only set to true', () => {
+    test('loads with the browse only checkbox checked', async ({ page }) => {
+      const conceptId = 'C1214470488-ASF'
+      const cmrHits = 0
+
+      await page.route('**/search/granules.json', (route) => {
+        const request = route.request()
+        const body = request.postData()
+        expect(body).toBe('browse_only=true&echo_collection_id=C1214470488-ASF&page_num=1&page_size=20')
```
Comment (trevorlang) [inline tests/e2e/paths/search/granules/granules.spec.js:370]: We generally have new lines to separate blocks of variables, expect blocks, and function calls. There are lots of occurrences of this in this file. Make sure all the spacing is consistent please :)

**Review by trevorlang — CHANGES_REQUESTED** [review]
I found a good bit of formatting inconsistencies that we should clean up but nothing major. I am a little confused how we ended up with 100k+ mocks in this PR if all we were doing is converting an already existing test. Maybe we can sync up tomorrow and you can fill me in on that.

File: tests/e2e/paths/search/granules/granules.spec.js (hunk 15 above)
Comment (dpesall) [reply tests/e2e/paths/search/granules/granules.spec.js:492]: I wondered the same thing, but this was the functionality in the previous test file I converted it from. I did console log both parts of the conditional in a few of these tests to see if they were both getting hit and it seemed like both were getting hit at times.

**Comment by dpesall** [comment]
> I found a good bit of formatting inconsistencies that we should clean up but nothing major. I am a little confused how we ended up with 100k+ mocks in this PR if all we were doing is converting an already existing test. Maybe we can sync up tomorrow and you can fill me in on that.

Happy to tag up tomorrow on this. I didn't create any new mocks for this test rewrite, I only reused the ones being utilized by the cypress test file. I'll definite work out all the spacing inconsistencies in the json files. Unless they somehow got changed by moving directories, they've been like that the whole time, so I'll go through and make sure they are all conforming.

**Comment by dpesall** [comment]
Added newlines to the end of all the files, and changed the spacing to be consistent across all files.

File: tests/e2e/paths/search/granules/granules.spec.js (hunk 16 above)
Comment (dpesall) [reply tests/e2e/paths/search/granules/granules.spec.js:385]: Fixed

**Review by trevorlang — APPROVED** [review]
